### ISSUE TITLE
Emformer pipeline demo results.

### ISSUE THREAD

**Issue by dykim3** [pytorch/audio#2514]
### 🚀 The feature

Hi. I'm testing Emformer model with pipeline_demo.py after 15 epoch.

after that, I got a few questions..

1. Why do you set memory to zero in default? The implement code doesn't use memory bank in training. I've thought it is helping the chunk to remember past utterances. 
2. What is the boundary effect? the original paper says left context, right context is concatenated with center context block because of boundary effect.
3. Is segment_length 16 size equal to 0.016 * 16 sec? (0.016 means the hop_length)
4. Is there any way to improve streaming accuracy by change train parameters? As you can see below, the streaming and non-streaming result is really different.. 

```  
Epoch 15: 100%|██████████| 98067/98067 [21:47:53<00:00, 968.80s/it, loss=8.62, v_num=20]Epoch 15, global step 1567776: 'Losses/val_loss' was not in top 1
Epoch 15, global step 1567776: 'Losses/train_loss' reached 6.10796 (best 6.10796), saving model to '/mnt/audio/examples/asr/emformer_rnnt/experiments/checkpoints/Emformer-epoch=15-train_loss=0.00.ckpt' as top 1
```

```
2022-06-28 15:33:23 streaming : 경기도 용인 씨 처ᅥ인 고 강가ᅡ 쪼ᅩᆷ 동백 ᄅ첨을려ᅥᆻ어그니ᅡᅡᅡᅡᅡᅡᅡ 이ᄅᆫ 습 가ᅳᆷᄎᆼ ᄋ리ᄅᆯ 저ᅳᆫ 삼백 일씨ᅩ 
2022-06-28 15:35:24 non-streaming : 경기도 용인시 처인구 장가 또 함 십 팔 자식을 동백 로지스 물류였으니까 일을 십 삼 층 일을 전 삼백 일십육고
2022-06-28 15:44:22 streaming : 서ᅮᆯ 서ᅡ초 서ᄎ워ᅭᄋᆯ 저ᅧᆫ 아ᅳᆫ데 사ᅡᆷ 식 가ᅳᆨ 가ᅵᅵᅵ 혀ᅳᆫᄃᆯᄋ 구층 이ᅧᅥᆫ 구ᄋᆨ이ᅩᅩ
2022-06-28 15:46:18 non-streaming : 서울특별시 서초구 서초동일천삼백삼십팔십현댁십온 일십구층일천구백이호
2022-06-28 15:49:02 streaming : ᆷ 로 간ᄇᆨ 즈ᅳᆯ소ᅩᆨ요
2022-06-28 15:49:45 non-streaming : 난 못 쓰는 로 간백 즐시고 낄
2022-06-28 15:53:37 streaming :  겨ᅮᅡᆯ 가ᅡᆼ나ᅩᅮ 토ᅩᅩᆨ도ᅮ 
2022-06-28 15:54:45 non-streaming : 아 원래 강남고 토곡동 상황에 사십 팔 학식이
2022-06-28 15:57:36 streaming :  가ᅩᆯ 가ᅡᆼ나ᅮᆫ 로ᅩᅥ 혹십 사ᅡᆷ
2022-06-28 15:58:24 non-streaming : 사 월 강남 오 일 향대천무 뺑이 혹시 삼
2022-06-28 16:06:29 streaming : 양 저ᅥᆫ고ᅩ 뭐 너 오ᅢ 호시ᅵᆨ 바ᅳᆯ 아ᅵᄌᆯ 사ᅡᆼ보ᅩ 바ᅡᅡ페ᅳᆯ리ᅵᄉᆺ스ᅳᅥ 이ᅵᆸ치ᅵᆯ 너ᅮᅮᆼ 이 저ᅥᆫ 사ᄇ배ᅢᆨ이고ᅵᅵᆷ유ᅥᅥ
2022-06-28 16:08:28 non-streaming : 사 월 명씩 양 전고 뭐 넘 오백 오십 팔 아시치를 상복 바깥에 알리스 이 차 바트 이백 일십 칠 넘 이천 사백 일십육호
2022-06-28 16:16:40 streaming :  ᄉᆯ허ᅥᆯ터 ᄋᆹ이이 더 먹은 거 더 더ᅮᅮᆫ 파ᅡ크ᅳᆸ 일 가ᅡᆷᄃᆫ지ᄋᆨ하ᅵ 그ᅳᅳᅳᅳ ᄋᆯ 시ᅳᆸ 요 이ᅥᄉᆫ 아ᅵ다ᅵ ᄋᆯ 시ᅵᆸ이요
2022-06-28 16:18:41 non-streaming : 서울특별시 도봉구 덕원동 육백사십사동파크를 간단 지약할 때 이백십 일십 육 넘 일 천 사백 일십 이호
2022-06-28 16:28:00 streaming :  겨ᅡᆼ 이더ᅥᆼ배ᅡᆨ 쪼ᅳᆫ 대대로 반 ᄀᆯ 사ᅵ 일 시ᅵᆫᄃᆼᄃᆯᄅᆺ대 캐ᅵᄉᆯ 일ᄃᆫ지ᅵ 일ᄇᆨ 이 유시ᅩᅩ 누 저ᅳᆫ ᄇᆷᄇᆨ 일 습고ᅩᅩ
2022-06-28 16:30:17 non-streaming : 경기도 연백 쪽전대로 오백 이십 칠 번 길 사십 일 신동백 롯데 캐슬 얼 일단 이제 일 백 일십 오 넘 이천 반백 일십 곡
```

thanks!

### Motivation, pitch

.

**Comment by thanhtvt** [comment]
About your first question, I think it is already being stated in the [paper](https://arxiv.org/pdf/2010.10759.pdf) (section **3.1.1**)

> For low latency experiments where the segment size is small, memory bank information largely overlaps with direct left context. Therefore, we set the memory bank length to 0.

**Comment by hwangjeff** [comment]
Hi @dykim3 — sorry for the late reply, and thanks for the questions.

1. @thanhtvt is exactly right — this particular model configuration is meant to optimize for low-latency operation, and so the memory length is set to 0.
2. The boundary effect refers to the impact of having sharp data cutoffs at the left and right boundaries of the input segment (center context), which is remedied by including additional data beyond the boundaries as left and right contexts.
3. In this case, a segment length of 16 frames translates to $\frac{16 \\, \text{frames}/\text{segment} \\,*\\, 160 \\, \text{samples} / \text{frame}}{16000 \\, \text{samples}/\text{s}} = 0.16 \\, \text{s} / \text{segment}$
4. There’s a tradeoff between latency and streaming accuracy. Specifically, if you make the input segment and left/right context longer, you’d expect the accuracy to improve, but at the expense of decoding speed. Depending on what segment and context sizes you land on, you may need to tune training parameters to achieve convergence.

**Comment by lakshmi-speak** [comment]
@hwangjeff In the emformer recipe, is the training mode and non-streaming forward pass still enforcing the `segment_length` through attention masking? Should the WER be similar in both modes?
I am noticing a large discrepancy in transcription quality with streaming inference and non-streaming forward pass on a custom dataset, would you know why that might be the case?

**Comment by lakshmi-speak** [comment]
> About your first question, I think it is already being stated in the [paper](https://arxiv.org/pdf/2010.10759.pdf) (section **3.1.1**)
> 
> > For low latency experiments where the segment size is small, memory bank information largely overlaps with direct left context. Therefore, we set the memory bank length to 0.

Isn't the memory more critical when the segment size is small? As in, the model has fewer context frames?
